Pegasus/include/Geometry.hpp

```diff
@@ -354,23 +312,78 @@ struct IntersectionCache<Box, Box> : IntersectionCacheBase
     double penetration = 0;
 };
 
+/** Data structure to store ray factors for Ray, Sphere collisions */
+struct RayIntersectionFactors
+{
+    double tMin;
+    double tMax;
+};
+
 /**
- * @brief Performs initial calculations and writes it into the cache object
- * 
- * Required to be called before any other proximity query function, otherwise results are undefined
- * @tparam ShapeA SimpleShape or SimpleShape derived object
- * @tparam ShapeB SimpleShape or SimpleShape derived object
- * @param[in] a pointer to the ShapeA type object
- * @param[in] b pointer to the ShapeB type object
- * @param[out] cacheBase pointer to the IntersectionCacheBase or IntersectionCacheBase derived object
+* @brief Calculates wether a ray and a sphere are intersecting
+* @param[in] raySphere vector from the ray to the sphere center
+* @param[in] sphereRadius radius of the sphere
+* @param[in] rayDirection normalized direction vector of the ray
+* @return @c true if there is intersection, @c false otherwise
+*/
+bool CalculateRaySphereIntersection(
+    glm::dvec3 const& raySphere, double sphereRadius, glm::dvec3 const& rayDirection
+);
+
+/**
+* @brief Calculates ray factors for the sphere intersection points
+*
+* Must be called if and only if the ray and the sphere are intersecting
+* @param[in] raySphere vector from the ray to the sphere center
+* @param[in] sphereRadius radius of the sphere
+* @param[in] rayDirection normalized direction vector of the ray
+* @return ray intersection factors
+*/
+RayIntersectionFactors CalculateRaySphereIntersectionFactors(
+    glm::dvec3 const& raySphere, double sphereRadius, glm::dvec3 const& rayDirection
+);
+
+/** Data structure for storing AABB using minimum and maximum points */
+struct AabbExtremalVertices
+{
+    glm::dvec3 minVertex;
+    glm::dvec3 maxVertex;
+};
+
+/**
+* @brief Calculates ray intersection factors for AABB, Ray collision
+* @param[in] boxMinPoint min point of AABB
+* @param[in] boxMaxPoint max point of AABB
+* @param[in] rayDirection normalized direction vector
+* @param[in] rayOrigin ray origin
+* @return in and out point factors
```

Comment:
is there a difference between `ray intersection factors` and `in and out point factors`?